src/murfey/util/lif.py

```diff
@@ -192,28 +184,164 @@ def rescale_to_bit_depth(
     # Rescale (DIVIDE BEFORE MULTIPLY)
     arr = (arr / (2**bit_init - 1)) * (2**bit_final - 1)
 
-    # This step probably not needed anymore
-    # Overwrite pixels that exceed channel bit depth
-    # arr[arr >= (2**bit_final - 1)] = 2**bit_final - 1
-
     # Change to correct unsigned integer type
     arr = change_bit_depth(arr, bit_final)
 
     return arr, bit_final
 
 
+def process_image_stack(
+    file: Path,
+    scene_num: int,
+    metadata: ET.Element,
+    save_dir: Path,
+):
+    """
+    Takes the LIF file and its corresponding metadata and loads the relevant sub-stack,
+    with each channel as its own array. Rescales their intensity values to utilise the
+    whole channel, scales them down to 8-bit, then saves each each array as a separate
+    TIFF image stack.
+    """
+
+    # Load LIF file
+    image = LifFile(str(file)).get_image(scene_num)
+
+    # Get name of sub-image
+    img_name = metadata.attrib["Name"]  # Get sub-image name
+    logger.info(f"Examining {img_name}")
+
+    # Create save dirs for TIFF files and their metadata
+    img_dir = save_dir / img_name
+    img_xml_dir = img_dir / "metadata"
+    for folder in [img_dir, img_xml_dir]:
+        if not folder.exists():
+            folder.mkdir(parents=True)
+            logger.info(f"Created {folder}")
+        else:
+            logger.info(f"{folder} already exists")
+
+    # Save image stack XML metadata (all channels together)
+    img_xml_file = img_xml_dir / (img_name + ".xml")
+    metadata_tree = ET.ElementTree(metadata)
+    ET.indent(metadata_tree, "  ")
+    metadata_tree.write(img_xml_file, encoding="utf-8")
+    logger.info(f"Image stack metadata saved to {img_xml_file}")
+
+    # Load relevant metadata (channels, dimensions, timestamps etc.)
+    channels = metadata.findall(
+        "Data/Image/ImageDescription/Channels/ChannelDescription"
+    )
+    # Might be useful in the future
+    # timestamps = elem.find("Data/Image/TimeStampList")
+    # dimensions = elem.findall(
+    #     "Data/Image/ImageDescription/Dimensions/DimensionDescription"
+    # )
+
+    # Parijat wants the images in 8-bit; scale down from 16-bit
+    # Save channels as individual TIFFs
+    for c in range(len(list(image.get_iter_c()))):
+        # Get color
+        color = channels[c].attrib["LUTName"]
+        logger.info(f"Examining the {color.lower()} channel")
+
+        # Extract image data to array
+        logger.info("Loading image stack")
+        arr: np.ndarray = []  # Array to store frames in
```

Comment:
This isn't an `np.ndarray` yet, it's a list

Reply:
I'm surprised mypy doesn't complain about it because `np.ndarray` has no `append` method. You could make it a numpy array at initialisation and use `np.append` inside the for loop